Add emulator support

ckaraneen opened the pull request:
- Add Emulator object that emulates bpod hardware behavior and reference
  it in the BpodBase class
- Modify bpod classes to accommodate emulator functionality
- Add emulator examples to showcase its usage

oraby commented on examples/emulator_examples/10_state_example.py:
```diff
@@ -0,0 +1,55 @@
+    state_name='State10',
+    state_timer=1,
+    state_change_conditions={Bpod.Events.Tup: 'exit'},
+    output_actions=[])
```
This can be simplified to:
```suggestion
for i in range(9):
   sma.add_state(
       state_name='State{}'.format(i),
       state_timer=1,
       state_change_conditions={Bpod.Events.Tup: 'State{}'.format(i+1)},
       output_actions=[])
sma.add_state(
    state_name='State10',
    state_timer=1,
    state_change_conditions={Bpod.Events.Tup: 'exit'},
    output_actions=[])
```

oraby commented on examples/emulator_examples/10_state_example_2.py:
```diff
@@ -0,0 +1,55 @@
+from pybpodapi.protocol import Bpod, StateMachine
```
As far as I can tell, this example will block as `Bpod.Events.GlobalTimer1_Start` will never fire. If that's the intent, then it should be documented.
I don't know if this project has a standard format for documenting module, but if not then you can add module docstring as in [this example](https://stackoverflow.com/a/23450896/11996983)

oraby replied:
The part I don't like about blocking examples is that it won't play nicely with CI tests.

ckaraneen replied:
I'll remove blocking examples.

oraby commented on examples/emulator_examples/10_state_example_2.py:
```diff
@@ -0,0 +1,55 @@
+from pybpodapi.protocol import Bpod, StateMachine
+my_bpod = Bpod(emulator_mode=True)
+sma = StateMachine(my_bpod)
+sma.add_state(
```
Also this example can be simplified as the one above.

oraby commented on examples/emulator_examples/10_state_example_2_unblock.py:
```diff
@@ -0,0 +1,59 @@
+# trigger. Link to LED of port 2.
+sma.set_global_timer(timer_id=1, timer_duration=0, on_set_delay=1.5,
+                     channel=Bpod.OutputChannels.PWM2, on_message=255)
+sma.add_state(
```
States 1 - > 8 can also be simplified here.

oraby replied:
If these tests will be part of a CI (which we need pybpod-api developers on), then it's probably better to decrease the timer duration and the number of states, otherwise we are unnecessarily increasing each test time (and CI bots might be configured to kill and refuse tests that take too long).

ckaraneen replied:
> States 1 - > 8 can also be simplified here.

Right.

> If these tests will be part of a CI (which we need pybpod-api developers on), then it's probably better to decrease the timer duration and the number of states, otherwise we are unnecessarily increasing each test time (and CI bots might be configured to kill and refuse tests that take too long).

See my comment regarding CI.

oraby commented on examples/emulator_examples/add_trial_events.py:
```diff
@@ -0,0 +1,58 @@
+import secrets
```
The original [example](https://github.com/pybpod/pybpod-api/blob/master/examples/state_machine_examples/add_trial_events.py#L12) uses `random`. I think also that `random` makes more sense here.

ckaraneen replied:
I had random but the codacy (github's code quality bot) warned that it's insecure. Although we don't care about his within our context, I added `secrets` for the sake of no unnecessary warnings. Just to be 'clean', I am more inclined to committing `random` -> `secrets` in all examples that contain it, if you agree.

ckaraneen replied:
Maybe we should change random -> secrets in a future PR in order to solve the Codacy security error:

> Standard pseudo-random generators are not suitable for security/cryptographic purposes.

oraby replied:
It's up to the pybpod developers, but IMO if we aren't doing crypto then it's misleading (and probably more computationaly intensive, but not the issue here) to use `secrest` rather than `random`. I would indeed prefer to leave them all as `random`

ckaraneen replied:
Will fix.

oraby commented on examples/emulator_examples/__init__.py:
```diff
@@ -0,0 +1,2 @@
+# !/usr/bin/python3
+# -*- coding: utf-8 -*-
```
I don't know where to add this comment about examples/tests in general so I'll add it here.
As far as I can tell, we can categorize tests into 2 categories:
1. Blocking vs. Non-Blocking tests
2. Tests that already reuse https://github.com/pybpod/pybpod-api/tree/master/examples/state_machine_examples and other new tests.

Possible improvements are:
1. For blocking tests, the user should have a warning about it, or a way to terminate it. Also they don't play nicely with CI. I wonder if we should remove them and keep just their manual-override version.
2. For non-blocking tests that are copied from the `state_machine_examples` folder, I wonder if we can find a way to include the code from the other modules, rather than rewrite, and inject the emulator before the run. In such way, only one version of the code is maintained and they would never run out of sync. However using the current design, i.e, using `Bpod(emulator_mode)`, would require a workaround.
3. For non-blocking tests generally, we should check and assert the resulting run events using [`current_trial`](https://pybpod-api.readthedocs.io/en/latest/pybpodapi/session.html#pybpodapi.session.Session.current_trial) or `trials`.  This can also serve as the basis for unit-tests for any future changes in the whole project.

I don't know about the pybpod-api developers, but IMO we shouldn't block this PR to wait for all changes, rather these can be considered (if they are willing) more of a future road-map.

oraby replied:
On a second thought, I think we should have at least one example in this PR where we check the resulting `current_trial` output, otherwise there is nothing here that tests that part.

ckaraneen replied:
> 1. For blocking tests, the user should have a warning about it, or a way to terminate it. Also they don't play nicely with CI. I wonder if we should remove them and keep just their manual-override version.

Will remove the blocking tests.

> 2. For non-blocking tests that are copied from the state_machine_examples folder, I wonder if we can find a way to include the code from the other modules, rather than rewrite, and inject the emulator before the run. In such way, only one version of the code is maintained and they would never run out of sync. However using the current design, i.e, using Bpod(emulator_mode), would require a workaround

Sounds reasonable. Let me return to this one once I've cleaned up the examples.

> On a second thought, I think we should have at least one example in this PR where we check the resulting current_trial output, otherwise there is nothing here that tests that part.

Since these were intended to be high-level behavioral tests for the user and not unit tests, I think we should avoid unit-test-like checks (e.g. assert) for these. Unit tests should be created as part of another PR.

oraby replied:
> Since these were intended to be high-level behavioral tests for the user and not unit tests, I think we should avoid unit-test-like checks (e.g. assert) for these. Unit tests should be created as part of another PR.

True, however the core value of the emulator is to produce an output raw-events/raw-states as if it works in the real life. In such way, the user can test the update of trial settings according to the animal action (e.g as in `updateCustomFields()` in `Mouse2AFC`).
I'm not proposing that we make sweep test for the resulting raw-events/states in this PR, but at least one test should check that we get any sensible output.

ckaraneen replied:
Sure, I will create one.

oraby commented on examples/emulator_examples/add_trial_events_2.py:
```diff
@@ -0,0 +1,81 @@
+import secrets
```
`random` as well.

oraby replied:
This example would also block.

oraby commented on pybpodapi/settings.py:
```diff
@@ -1,15 +1,16 @@
+import logging
 
-PYBPOD_API_LOG_LEVEL = None
+PYBPOD_API_LOG_LEVEL = logging.DEBUG
```
All the changes in this file are not related to this PR, and should be discarded accordingly.

oraby commented on pybpodapi/bpod/emulator/__init__.py:
```diff
@@ -0,0 +1 @@
+from .emulator import Emulator
```
I think most files here ends with new empty line.

oraby commented on pybpodapi/bpod/bpod_com_protocol_modules.py:
```diff
@@ -123,11 +123,9 @@ def _bpodcom_get_modules_info(self, hardware):
             [ord("%")] + n_serial_events_array + [n_soft_codes]
         )
-
         self._arcom.write_array(bytes2send)
```
These changes are unrelated to this commit.

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -59,42 +61,61 @@ class ChannelNames(ChannelName):
-    def __init__(self, serial_port=None, sync_channel=None, sync_mode=None, net_port=None):
+    def __init__(self, serial_port=None, sync_channel=None, sync_mode=None, net_port=None, emulator_mode=False):
         self._session = self.create_session()
-
```
This line change is unrelated.

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -59,42 +61,61 @@ class ChannelNames(ChannelName):
+        self.session += SessionInfo(self.session.INFO_SETUP_NAME,
+                                    settings.PYBPOD_SETUP)
+        self.session += SessionInfo(self.session.INFO_BPODGUI_VERSION,
+                                    settings.PYBPOD_BPODGUI_VERSION)
```
Also the changes and the few below are unrelated to this PR. If you want to include them, then add them to as a separate formatting commit.

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -222,8 +251,9 @@ def send_state_machine(self, sma, run_asap=None):
-        if not self.bpod_com_ready:
-            raise Exception('Bpod connection is closed')
+        if not self._emulator.enabled:
+            if not self.bpod_com_ready:
```
Why not just:
```suggestion
        if not self._emulator.enabled and not self.bpod_com_ready:
```

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -259,29 +294,38 @@ def run_state_machine(self, sma):
-            raise Exception('Bpod connection is closed')
+        if not self._emulator.enabled:
+            if not self.bpod_com_ready:
+                raise Exception('Bpod connection is closed')
```
Same:
```suggestion
        if not self._emulator.enabled and not self.bpod_com_ready:
             raise Exception('Bpod connection is closed')
```

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -259,29 +294,38 @@ def run_state_machine(self, sma):
 
-        self.trial_timestamps = []  # Store the trial timestamps in case bpod is using live_timestamps
+        # Store the trial timestamps in case bpod is using live_timestamps
+        self.trial_timestamps = []
```
Also changes are not related to this commit.

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -291,14 +335,21 @@ def run_state_machine(self, sma):
+                sma.current_state)
+            # BpodSystem.RefreshGUI;
         sma.is_running = True
+
```
Unrelated extra line

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -291,14 +335,21 @@ def run_state_machine(self, sma):
+            self._emulator.initialize()
+            self._emulator.mirror_state(
+                sma.current_state)
+            # BpodSystem.RefreshGUI;
```
Do you need the commented line? If not, then we should remove it, if it' necessary for some cases, then add a second comment on the right of it stating when it should be uncommented.

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -259,29 +294,38 @@ def run_state_machine(self, sma):
+                        'Error: The last state machine sent was not acknowledged by the Bpod device.', self)
+            self.trial_start_timestamp = self._bpodcom_get_trial_timestamp_start()
+            if self.bpod_start_timestamp is None:
+                self.bpod_start_timestamp = self.trial_start_timestamp
```
The overlap between the two branches are quite common that it make sense to unify it more:
```python
        if not self._emulator.enabled:
           self._bpodcom_run_state_machine()
        if self._new_sma_sent:
            if self._emulator_enabled:
               self._emulator.set_state_machine(sma) # We can move code as part of _bpodcom_state_machine_installation_status() function
               self._new_sma_sent = False
            elif self._bpodcom_state_machine_installation_status(): # Not emulator
               self._new_sma_sent = False
            else:
                raise BpodErrorException('Error: The last state machine sent was not acknowledged by the Bpod device.', self)
            self.trial_start_timestamp = self._bpodcom_get_trial_timestamp_start() # Probably this require handling for emulator
            if self.bpod_start_timestamp is None:
                self.bpod_start_timestamp = self.trial_start_timestamp
```

I don't if that was bug or not, but in your code `self.trial_start_timestamp` was not set in the emulator branch.

ckaraneen replied:
> I don't if that was bug or not, but in your code self.trial_start_timestamp was not set in the emulator branch.

Yep, it's a bug.

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -291,14 +335,21 @@ def run_state_machine(self, sma):
                 if inline is not None:
-                    interrupt_task, kill_task = self.handle_inline(inline, sma)
+                    interrupt_task, kill_task = self.handle_inline(
+                        inline, sma)
```
Also unrelated changes

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -307,19 +358,30 @@ def run_state_machine(self, sma):
                     inline = inline.decode().strip()
-                    interrupt_task, kill_task = self.handle_inline(inline, sma)
+                    interrupt_task, kill_task = self.handle_inline(
+                        inline, sma)
```
Same here

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -400,7 +465,8 @@ def reset_serial_messages(self):
         if not response:
-            raise BpodErrorException('Error: Failed to reset serial message library.')
+            raise BpodErrorException(
+                'Error: Failed to reset serial message library.')
```
The 4 changes above are unrelated as well.

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -411,19 +477,42 @@ def softcode_handler_function(self, data):
 
     def echo_softcode(self, softcode):
-        return self._bpodcom_echo_softcode(softcode)
+        self._bpodcom_echo_softcode(softcode)
```
I think this is better and semi-related to readability of this PR, but I think it still make sense to separate this change and the similar no-return changes into a separate commit.

ckaraneen replied:
I agree.

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -411,19 +477,42 @@ def softcode_handler_function(self, data):
     def trigger_softcode(self, softcode):
-        return self._bpodcom_send_softcode(softcode)
+        if self._emulator.enabled:
+            self.softcode_handler_function(value)
```
Why is this not part of the emulator if it's just used with it?

ckaraneen replied:
It's also used in `__process_opcode()`.

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -411,19 +477,42 @@ def softcode_handler_function(self, data):
+
+    def trigger_serial(self, channel_number, value):
+        if self._emulator.enabled:
+            self._emulator.override_output_state(channel_number, value)
```
would it be better to call that function `_emulator.serial_override(channel_number, value)`?

ckaraneen replied:
I agree.

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -439,6 +528,47 @@ def load_message(self, module_index, msg):
+                if debug_message is not None:
+                    logger.debug(debug_message)
+                current_trial.states.append(
+                    sma.current_state)
```
wouldn't that fit in one line?

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -439,6 +528,47 @@ def load_message(self, module_index, msg):
+        current_state = sma.current_state
+        if is_state_timer_matrix:
+            this_state_timer_state = transition_matrix[current_state]
+            if event_id == sma.hardware.channels.events_positions.Tup:
+                if this_state_timer_state != current_state:
```
these can be combined in a single if-condition, right?

ckaraneen replied:
For the sake of readability it's better if I do:
```python
is_event_id_tup = event_id == sma.hardware.channels.events_positions.Tup
if is_event_id_tup and this_state_timer_state != current_state:
```

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -439,6 +528,47 @@ def load_message(self, module_index, msg):
+                                  f'for transition: state '
+                                  f'{sma.current_state} -> '
+                                  f'{transition_state}'))
+        return new_state_set
```
I think it'd increase readability to move `new_state_set` outside the nested function and set it in the if-else branches:
```python
        def set_sma_current_state(new_state):
            previous_state = sma.current_state
            if sma.use_255_back_signal and new_state == 255:
                sma.current_state = current_trial.states[-2]
            else:
                sma.current_state = new_state
            logger.debug(('Transition occured: '
                          f'state {previous_state} -> {sma.current_state}'))
            if not math.isnan(sma.current_state):
                if debug_message is not None:
                    logger.debug(debug_message)
                current_trial.states.append(
                    sma.current_state)
                state_change_indexes.append(
                    len(current_trial.events_occurrences) - 1)
        current_state = sma.current_state
        new_state_set = False
        if is_state_timer_matrix:
            this_state_timer_state = transition_matrix[current_state]
            if event_id == sma.hardware.channels.events_positions.Tup:
                if this_state_timer_state != current_state:
                    set_sma_current_state(this_state_timer_state)
                    new_state_set = True
        else:
            for transition_event_code, transition_state in transition_matrix[
                    current_state]:
                if transition_event_code == event_id:
                    set_sma_current_state(transition_state)
                    new_state_set = True
                else:
                    logger.debug((f'Event {transition_event_code} required '
                                  f'for transition: state '
                                  f'{sma.current_state} -> '
                                  f'{transition_state}'))
        return new_state_set
```

ckaraneen replied:
I agree.

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -474,83 +609,59 @@ def __process_opcode(self, sma, opcode, data, state_change_indexes):
+                self._emulator.mirror_events(current_events)
             if transition_event_found and not math.isnan(sma.current_state):
+                if self._emulator.enabled:
+                    self._emulator.mirror_state(sma.current_state)
+                    self._emulator.set_state_machine(sma)
+                    self._emulator.reset_current_state()
+                    self._emulator.set_state_start_time(
+                        self._emulator.current_time)
+                    self._emulator.set_global_timer_end_time()
+                    self._emulator.cancel_global_timers()
+                    self._emulator.reset_global_counter_counts()
+                    self._emulator.set_softcode(
+                        self.hardware.channels.events_positions.output_USB)
```
This code looks to like it belong inside the emulator rather than in here. If the original Matlab code keeps that way, then it should be kept that way, otherwise it should be wrapped inside the emulator in a function.

ckaraneen replied:
The original Matlab code keeps that way.

oraby replied:
hmm, looking at this again, I would say that Matlab code left it this way because it doesn't utilize classes, but here I think it might make sense to wrap it all in a reset function as originally suggested. What do you think?

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -474,83 +609,59 @@ def __process_opcode(self, sma, opcode, data, state_change_indexes):
-                                transition_event_found = True
-
+                    logger.debug("Current state: %s", sma.current_state)
+                    transition_matrices = {
+                        'input': sma.input_matrix,
+                        'state_timer': sma.state_timer_matrix,
+                        'global_timers_start': sma.global_timers.start_matrix,
+                        'global_timers_end': sma.global_timers.end_matrix,
+                        'global_counters': sma.global_counters.matrix,
+                        'conditions': sma.conditions.matrix,
+                    }
+                    for transition_matrix_name, transition_matrix in \
+                            transition_matrices.items():
+                        transition_event_found = \
+                            self.__transition_to_new_state(
+                                sma,
+                                event_id,
+                                transition_matrix,
+                                current_trial,
+                                state_change_indexes,
+                                is_state_timer_matrix=(transition_matrix_name ==
+                                                       'state_timer'),
+                                debug_message="Adding {} matrix states".format(
+                                    transition_matrix_name))
+                        if transition_event_found:
+                            break
```
I think this is a very nice refactoring, but it's unrelated directly (as far as I can tell) to the emulator. What I would propose is to add a commit *before* the emulator which implements this change. This would make reverse change modular and would facilitate the review of the change.

ckaraneen replied:
I agree.

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -560,39 +671,89 @@ def __update_timestamps(self, sma, state_change_indexes):
         current_trial.state_timestamps += timestamps[-1:]
 
+    def manual_override(self, channel_type, channel_name, channel_number, value):
+        """
+        Manually override a Bpod channel
+
+        :param ChannelType channel_type: channel type input or output
+        :param ChannelName channel_name: channel name like PWM, Valve, etc.
+        :param channel_number:
+        :param int value: value to write on channel
+        """
+        if channel_type == ChannelType.INPUT:
+            input_channel_name = channel_name + str(channel_number)
+            try:
+                channel_number = self.hardware.channels.input_channel_names.index(
+                    input_channel_name)
+                self.trigger_input(channel_number, value)
+            except:
+                raise BpodErrorException(
+                    'Error using manual_override: {name} is not a valid channel name.'.format(name=channel_name))
+
+        elif channel_type == ChannelType.OUTPUT:
+            if channel_name == 'SoftCode':
+                self.trigger_softcode(value)
+            elif channel_name == 'Serial':
+                self.trigger_serial(channel_number, value)
+            else:
+                output_channel_name = channel_name + \
+                    str(channel_number)
+                try:
+                    channel_number = self.hardware.channels.output_channel_names.index(
+                        output_channel_name)
+                    self.trigger_output(channel_number, value)
+                except:
+                    raise BpodErrorException('Error using manual_override: {name} is not a valid channel name.'.format(
+                        name=output_channel_name))
+        else:
+            raise BpodErrorException(
+                'Error using manualOverride: first argument must be "Input" or "Output".')
+
```
Is there a reason for moving this function from the `bpod_com_protocol` module to here? I failed to spot a necessity for it.

This is the diff between the original location and this change here:
```diff
         if channel_type == ChannelType.INPUT:
             input_channel_name = channel_name + str(channel_number)
-            channel_number = self.hardware.channels.input_channel_names.index(input_channel_name)
             try:
-                self._bpodcom_override_input_state(channel_number, value)
+                channel_number = self.hardware.channels.input_channel_names.index(
+                    input_channel_name)
+                self.trigger_input(channel_number, value)
             except:
                 raise BpodErrorException(
                     'Error using manual_override: {name} is not a valid channel name.'.format(name=channel_name))

         elif channel_type == ChannelType.OUTPUT:
-            if channel_name == 'Serial':
-                self._bpodcom_send_byte_to_hardware_serial(channel_number, value)
-
+            if channel_name == 'SoftCode':
+                self.trigger_softcode(value)
+            elif channel_name == 'Serial':
+                self.trigger_serial(channel_number, value)
             else:
+                output_channel_name = channel_name + \
+                    str(channel_number)
                 try:
-                    output_channel_name = channel_name + str(channel_number)
-                    channel_number = self.hardware.channels.output_channel_names.index(output_channel_name)
-                    self._bpodcom_override_digital_hardware_state(channel_number, value)
+                    channel_number = self.hardware.channels.output_channel_names.index(
+                        output_channel_name)
+                    self.trigger_output(channel_number, value)
                 except:
                     raise BpodErrorException('Error using manual_override: {name} is not a valid channel name.'.format(
                         name=output_channel_name))
         else:
-            raise BpodErrorException('Error using manualOverride: first argument must be "Input" or "Output".')
+            raise BpodErrorException(
+                'Error using manualOverride: first argument must be "Input" or "Output".')

```

ckaraneen replied:
> Is there a reason for moving this function from the bpod_com_protocol module to here? I failed to spot a necessity for it.

Since I add emulator logic, the function no longer contains only com logic. Hence, the move.

oraby replied:
But nothing the code there is an emulator code, it's mainly calling base class functions. In such case, it might more sense to leave it in the child class.
Also, some of the changes here are unrelated, e.g: moving code inside the try/catch block.

ckaraneen replied:
The trigger_* functions that this function calls check whether the emulator mode is enabled and dispatch an action accordingly. I made the move since it no longer **only** calls 'com' stuff, but also emulator stuff (through the trigger functions).

oraby replied:
my point is that the function bodies are implementation details that are transparent to the children classes, and as such, it's a 'legal' action to call the parent method.

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -560,39 +671,89 @@ def __update_timestamps(self, sma, state_change_indexes):
         current_trial.event_timestamps = timestamps
-        current_trial.state_timestamps += [timestamps[i] for i in state_change_indexes]
+        current_trial.state_timestamps += [timestamps[i]
+                                           for i in state_change_indexes]
```
Same unrelated changes

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -560,39 +671,89 @@ def __update_timestamps(self, sma, state_change_indexes):
+                trial_end_timestamp - self.trial_start_timestamp,
+                start_time=self.trial_start_timestamp,
+                end_time=trial_end_timestamp
+            )
```
Isn't that part common to both the emulator and the real hardware?

ckaraneen replied:
Correct. Fixed.

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -59,42 +61,61 @@ class ChannelNames(ChannelName):
         self.sync_mode = sync_mode if sync_mode is not None else settings.PYBPOD_SYNC_MODE
         self.net_port = net_port if net_port is not None else settings.PYBPOD_NET_PORT
-        self._hardware = Hardware()    # type: Hardware
+        self._emulator = Emulator(emulator_mode)
```
`net_port` argument is missing

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -59,42 +61,61 @@ class ChannelNames(ChannelName):
+            self._hardware = self._emulator.hardware
+            self.sock = self._emulator.sock
+            self.socketin = self._emulator.socketin
+            self.stdin = self._emulator.stdin
```
I'm not a fan of this part. On real hardware, these variables are initialized in [`open()`](https://github.com/pybpod/pybpod-api/pull/5/files#diff-c1bf0eb6016fc3d3327d1ee1ddb66991R192-R202), however whether the current design these variables are initialized here and are basically copied from the emulator code which replicates more or less the same code again as in the `open()` function. I think this makes the code more susceptible to run of sync as the code is distributed everywhere. I'd suggest instead to do all the initialization in the `open()` method in this class.

oraby replied:
Sample applies to `Hardware()` as far as I can tell.
If the emulator needs the same `Hardware()` object, then one possibility is to initialize it here and pass it to the `Emulator()` class.

oraby commented on pybpodapi/bpod/emulator/emulator.py:
```diff
@@ -0,0 +1,715 @@
+        if state is None:
+            self._state.reset_input()
+            self._state.reset_output()
+            # BpodSystem.RefreshGUI;
```
This needs to be a call to the gui-emulator-plugin if it's enabled, correct?. If so, then maybe it's worth adding a `#TODO: ....`

oraby reviewed (commented):
I've had a go through the code except for the actual emulator class, but I wanted to give some initial feedback first. Generally, I think the code is doing a good job where it basically satisfies what needs to be done, with the exception of few unrelated cosmetic changes. 
I have wished though if the emulator was implemented as an instance of a `BpodCOMProtocol()` class, however that's not currently possible as the current `Bpod` classes design uses inheritance when at many times composition would have been IMO a better choice.

One last thought: I wonder whether it'll be simpler to replace `emulator.enabled` with simply having `self._emulator` either as `Emulator()` or `None`.

ckaraneen commented:
> One last thought: I wonder whether it'll be simpler to replace emulator.enabled with simply having self._emulator either as Emulator() or None.

I agree.

**Regarding CI**:

If these tests will be part of CI, your aforementioned comments are true, and we 'll also have a problem with the threaded emulator examples, where I showcase the manual override functionality. But, these tests were intended to demonstrate emulator functionality in the spirit of the examples under the function_examples and state_machine_examples directories. All these are not unit tests.

All in all, I think that CI should **not** be configured to run these behavioral tests, but only unit tests, if they are added in the future (separate PR), which I totally recommend.

ckaraneen committed:
Add emulator support

- Add Emulator object for emulating bpod hardware behavior and
communicating outputs to emulator GUI server
- Modify bpod classes to accommodate emulator functionality
- Add emulator examples to showcase its usage

oraby commented on pybpodapi/settings.py:
```diff
@@ -19,6 +19,7 @@
 # TARGET_BPOD_FIRMWARE_VERSION = "15" # 0.8
 # TARGET_BPOD_FIRMWARE_VERSION = "17" # 0.9
 TARGET_BPOD_FIRMWARE_VERSION = "22"
+TARGET_BPOD_MACHINE_TYPE = 3
```
Since it only affects the emulator, would it make sense to indicate so, e.g:
`EMULATOR_BPOD_MACHINE_TYPE = 3`

oraby commented on examples/emulator_examples/global_counter_example_manual_override.py:
```diff
@@ -0,0 +1,56 @@
+                            channel_number=1, value=12)
+    time.sleep(1)
+    my_bpod.manual_override(Bpod.ChannelTypes.INPUT, 'Port',
+                            channel_number=1, value=12)
```
That can all be added in a loop:
```suggestion
    time.sleep(1)
    for i in range(5):
      time.sleep(1)
      my_bpod.manual_override(Bpod.ChannelTypes.INPUT, 'Port',
                              channel_number=1, value=12)
```

oraby commented on examples/emulator_settings.py:
```diff
@@ -0,0 +1,23 @@
+PYBPOD_API_ACCEPT_STDIN = False
+
+TARGET_BPOD_FIRMWARE_VERSION = "22"
+TARGET_BPOD_MACHINE_TYPE = 3
```
Same would be applied here if the name changed

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -163,18 +170,7 @@ def open(self):
-        # initialise the thread that will handle the stdin commands
-        self.stdin = NonBlockingStreamReader(sys.stdin) if settings.PYBPOD_API_ACCEPT_STDIN else None
-        #####################################################
+        self.__initialize_input_command_handler()
```
If this is emulator mode, then this function would be called twice, once here and once in https://github.com/pybpod/pybpod-api/pull/5/commits/e6527861225f3bf6cf192cf061e7297b39c6e748#diff-c1bf0eb6016fc3d3327d1ee1ddb66991R75

ckaraneen replied:
What do you mean? It's called once in \_\_init\_\_() and once inside open(), which is not called if emulator mode is enabled.

oraby replied:
yup, you are right

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -271,14 +272,25 @@ def run_state_machine(self, sma):
-            else:
-                raise BpodErrorException('Error: The last state machine sent was not acknowledged by the Bpod device.', self)
+            if self._emulator is None:
+                if not self._bpodcom_state_machine_installation_status():
```
These 2 if-conditions can be merged into one, right?

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -271,14 +272,25 @@ def run_state_machine(self, sma):
+                    raise BpodErrorException('Error: The last state machine sent was not acknowledged by the Bpod device.', self)
+            self._new_sma_sent = False
+        else:
+            if self._emulator is not None:
```
A bit of nit-picking, but wouldn't it be simpler to just have it as:
`if self._emulator:`
Also, rather than having an else and an if, we could just use an `elif self._emulator:`
I'm personally okay with leaving both as they are though

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -310,16 +322,26 @@ def run_state_machine(self, sma):
                     interrupt_task, kill_task = self.handle_inline(inline, sma)
             #####################################################
 
-            if self.data_available():
```
Unless I've missed something, `data_available()` is defined only in the child class, That's something that need fixing at some point, we probably should create an issue for it.

ckaraneen replied:
Keeping it in mind for a separate PR.

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -541,16 +604,67 @@ def __process_opcode(self, sma, opcode, data, state_change_indexes):
+            if self._emulator is not None:
+                self._emulator.mirror_events(current_events)
             if transition_event_found and not math.isnan(sma.current_state):
+                if self._emulator is not None:
```
Same suggestion here about just using `if self._emulator`

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -541,16 +604,67 @@ def __process_opcode(self, sma, opcode, data, state_change_indexes):
-                logger.debug("States indexes: %s", current_trial.states)
+                    # global counters matrix
+                    if not transition_event_found:
+                        for transition in sma.global_counters.matrix[sma.current_state]:
```
Why this for loop and the one below are needed now and weren't there before?. Was it a missing feature from pybpod or is it emulator specific?

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -541,16 +604,67 @@ def __process_opcode(self, sma, opcode, data, state_change_indexes):
+                    if not transition_event_found:
+                        for transition in sma.conditions.matrix[sma.current_state]:
+                            if transition[0] == event_id:
+
```
I know that [here](https://github.com/pybpod/pybpod-api/pull/5/commits/e6527861225f3bf6cf192cf061e7297b39c6e748#diff-c1bf0eb6016fc3d3327d1ee1ddb66991R595) there was en empty line inside the for loop, but [here](https://github.com/pybpod/pybpod-api/pull/5/commits/e6527861225f3bf6cf192cf061e7297b39c6e748#diff-c1bf0eb6016fc3d3327d1ee1ddb66991R579) and [here](https://github.com/pybpod/pybpod-api/pull/5/commits/e6527861225f3bf6cf192cf061e7297b39c6e748#diff-c1bf0eb6016fc3d3327d1ee1ddb66991R546) there aren't any. I think having an empty line here was accidental.
Same comment would apply [here](https://github.com/pybpod/pybpod-api/pull/5/commits/e6527861225f3bf6cf192cf061e7297b39c6e748#diff-c1bf0eb6016fc3d3327d1ee1ddb66991R611)

oraby commented on pybpodapi/bpod/emulator/state.py:
```diff
@@ -0,0 +1,50 @@
+class State(object):
+
+    # N_OUTPUTS_OFFSET = 3
+
```
Is it on purpose to leave an empty line before and after here?

oraby replied:
Also you need this commented out code?

oraby commented on pybpodapi/bpod/emulator/state.py:
```diff
@@ -0,0 +1,50 @@
+
+    def reset_output(self):
+        self._output = StateDict()  # dict(output code -> output value)
+        # In MATLAB: [0] * (hardware.n_outputs + self.N_OUTPUTS_OFFSET)
```
Shouldn't `reset()` mean set things to their initial value?, in such way, you expect objects to be ready to use in a clean state after resetting. I would argue then that assigning the value:
`self._input_type = hardware.inputs` should be part of the `reset()` function.
If your intention is to leave the implementierten as it is, then a better name would be `clear_input()` for example.
Same applies to the `reset_output()`.

oraby commented on pybpodapi/bpod/emulator/emulator.py:
```diff
@@ -0,0 +1,668 @@
+        self._setup_hardware_channels()
+        self._setup_hardware_event_types()
+
+    def _reset_global_variables(self):
```
Wouldn't a better name be: `_init_hardware_offsets()` or something similar, to avoid confusing the word global with the common usage of 'global variables'.

oraby commented on pybpodapi/bpod/emulator/emulator.py:
```diff
@@ -0,0 +1,668 @@
+        try:
+            return func(*args, **kwargs)
+        except AttributeError:
+            message = 'The emulator has not been initialized.'
```
How about: `The emulator initialize() has not been called.` to help the user more.

oraby commented on pybpodapi/bpod/emulator/emulator.py:
```diff
@@ -0,0 +1,668 @@
+        this_global_timer = self._output_value(
+            GLOBAL_TIMER_TRIG_OUTPUT_ACTION)
+        if this_global_timer != 0:
+            this_global_timer_idx = this_global_timer - 1
```
how about renaming it to `timer_idx` so you don't have to break so many lines onto mutilines?

oraby commented on pybpodapi/bpod/emulator/emulator.py:
```diff
@@ -0,0 +1,668 @@
+        self._assert_state_machine()
+        virtual_current_events = []
+        # MATLAB EQUIVALENT:
+        # virtual_current_events = [0] * N_VIRTUAL_CURRENT_EVENTS
```
Why we don't know here yet `N_VIRTUAL_CURRENT_EVENTS` in the same way that Matlab already knows about at this point?

oraby replied:
Just to clarify why Matlab is doing it in this way, the advantage of this approach in Matlab is to preallocate the array rather than keeping on expanding it. Pre-allocating is better as it avoids appending, where appending would allocate memory every few appends, and allocating memory in general is a bit computationally exhaustive.
I doubt in Python it makes much difference in performance, unless we pre-allocate the array as numpy array.
I apologize if I'm iterating already known knowledge.

ckaraneen replied:
I think it would be more wise to keep it simple and pythonic as possible for this PR. And maybe consider using numpy in this part but also elsewhere as part of a separate PR.

oraby commented on pybpodapi/bpod/emulator/emulator.py:
```diff
@@ -0,0 +1,668 @@
+            0
+            if highest_used_global_condition is None
+            else highest_used_global_condition + 1
+        )
```
How about:
```suggestion
        def getMaxUsed(sma_global):
            # Returned sma.*.get_max_index_used() can be None
            max_idx = sma_global.get_max_index_used()
            return 0 if max_idx is None else return max_idx+1
        highest_used_global_counter = getMaxUsed(sma.global_counters)
        highest_used_global_timer = getMaxUsed(sma.global_timers)
        highest_used_global_condition = getMaxUsed(sma.conditions.get_max)
```

oraby commented on pybpodapi/bpod/emulator/emulator.py:
```diff
@@ -0,0 +1,668 @@
+                else:
+                    if self._global_timers_triggered[i] and \
+                            not self._global_timers_active[i] and \
+                            self.current_time > self._global_timer_start[i]:
```
These can be combined as a single `elif`

oraby commented on pybpodapi/bpod/emulator/emulator.py:
```diff
@@ -0,0 +1,668 @@
+                        self._n_current_events += 1
+                        self._global_timers_active[i] = True
+                        self._global_timers_triggered[i] = False
+                        # self._set_global_timer_channel(i, 1)
```
Can you add a comment to this code mentioning why it's commented out

oraby commented on pybpodapi/bpod/emulator/emulator.py:
```diff
@@ -0,0 +1,668 @@
+"""
```
Generally speaking, and about this file in particular, I find that sticking to 80-column limit coulmn (which I'm fan of) but at the same time using a tab indentation of size 4, eats quite a lot of space, and forces breaking many lines into multi-lines. If the pybpod developers are fine with it, I'd recommend using 2 spaces tab instead of 4.

ckaraneen replied:
I'd advise against this actually since it is best to stick with the [PEP 8 style guide](https://www.python.org/dev/peps/pep-0008/#indentation).

oraby commented on pybpodapi/bpod/emulator/emulator.py:
```diff
@@ -0,0 +1,668 @@
+            # Evaluate condition transitions
+            for i in range(self.hardware.n_conditions):
+                sma_conditions = self._state_machine.conditions
+                condition_event_name = CONDITION_EVENT_NAME_PREFIX + str(i + 1)
```
For performance, you can initialize all the possible strings during the `reset_global_variables()`. This way, the strings aren't re-computed on every run.
It's a small optimization, so if you rather like, you can just add a TODO comment about it here instead.

oraby commented on pybpodapi/bpod/emulator/emulator.py:
```diff
@@ -0,0 +1,668 @@
+                        self._global_timers_triggered[i] = False
+                        # self._set_global_timer_channel(i, 1)
+            # Evaluate global counter transitions
+            for i in range(self.hardware.n_global_counters):
+                sma_global_counters = self._state_machine.global_counters
+                if sma_global_counters.attached_events[i] != \
+                        FINAL_GLOBAL_COUNTER_EVENT:
+                    if self._global_counter_counts[i] == \
+                            sma_global_counters.thresholds[i]:
+                        virtual_current_events.append(
+                            self._GLOBAL_COUNTER_OFFSET + i)
+                        self._n_current_events += 1
+                    if virtual_current_events and \
+                            virtual_current_events[0] == \
+                            sma_global_counters.attached_events[i]:
+                        self._global_counter_counts[i] += 1
```
I assume you are copying here Matlab code as is for the most part, if not, I wonder if we can utilize numpy to re-implement this part and the other for loops in some form like the following (utilizing vectorization and avoiding loops):
```python
idxs = np.where(sma_global_counters.attached_events[i] != FINAL_GLOBAL_COUNTER_EVENT)[0]
idxs1 = idxs[self._global_counter_counts[idxs] == sma_global_counters.thresholds[idxs]]
virtual_current_events.append(idxs1 + self._GLOBAL_COUNTER_OFFSET)
if len(virtual_current_events) and \
 (virtual_current_events[0] in sma_global_counters.attached_events[idxs]).any():
    self._global_counter_counts[i] += 1
```

ckaraneen replied:
Yes, I've copied that from Matlab. I think that incorporating numpy, etc. functionality should be part of a separate PR where we can refactor and rethink many parts that may be written more cleanly with such data science tools.

oraby commented on pybpodapi/bpod/emulator/emulator.py:
```diff
@@ -0,0 +1,668 @@
+                        target_state = destination_state
+                        break
+                if target_state is not None:
+                    if target_state != self._current_state:
```
How about:
`if target_state and target_state != self._current_state:`

oraby commented on pybpodapi/bpod/emulator/emulator.py:
```diff
@@ -0,0 +1,668 @@
+            # Evaluate state timer transitions
+            time_in_state = self.current_time - self._state_start_time
+            state_timer = self._state_machine.state_timers[self._current_state]
+            if (time_in_state > state_timer) and \
```
A bit of nit-picking, but is there a reason for the `(` and `)` here?

oraby commented on pybpodapi/bpod/emulator/emulator.py:
```diff
@@ -0,0 +1,668 @@
+                        [self.current_time - self._matrix_start_time] * \
+                        self._n_current_events
+                    self._n_events = total_n_events
+        else:
+            opcode = 2
+            data = self._softcode
+            self._softcode = 0
```
Is that part also a copy from Matlab code? If not, then we can refactor the if else to:
````python
if self._softcode != 0:
  opcode, data  = 2, self._softcode
  self._softcode = 0
  return opcode, data
## No else branch, the rest of the code simply continues here branchless
# self._softcode is 0
self.current_time = time.time()
self._n_current_events = 0
.
.
return opcode, data

ckaraneen replied:
Yes, it is.

oraby commented on pybpodapi/bpod/emulator/emulator.py:
```diff
@@ -0,0 +1,668 @@
+        # State timer matrix (state timer transitions)
+        state_timer_matrix = []
+        for i in range(sma.total_states_added):
+            state_timer_matrix += [
+                sma.total_states_added
+                if math.isnan(sma.state_timer_matrix[i])
+                else sma.state_timer_matrix[i]
+            ]
```
Since you are appending a single value, whose value differ from the branch, I think it'll be clearer to refactor it as follows:
```python
timer = sma.state_timer_matrix[i]
if math.isnan(timer)
  timer = sma.total_states_added
state_timer_matrix.append(timer)
```

Also, if utilize numpy here, I think we can replace the whole for loop:
```python
state_timer_matrix = sma.state_timer_matrix[i].copy()
np.nan_to_num(state_timer_matrix, copy=False, nan=sma.total_states_added) # Or I think: state_timer_matrix[np.isnan(state_timer_matrix)] = sma.total_states_added)
```

oraby replied:
Same first part of the comment about the parts below that uses `if math.isnan(...)`

oraby commented on pybpodapi/bpod/emulator/emulator.py:
```diff
@@ -0,0 +1,668 @@
+        for i in range(sma.total_states_added):
+            hw_state = sma.output_matrix[i]
+            hw_state = [
+                evt
+                for evt in hw_state
+                if evt[0] < global_timer_trigger_event_pos
```
Personally, I find list comprehensions are more redable if they are a one-liner. Maybe they can be done as:
```pyton
                pos = global_timer_trigger_event_pos
                hw_state = [evt for evt in hw_state if evt[0] < pos]
```

oraby commented on pybpodapi/bpod/emulator/emulator.py:
```diff
@@ -0,0 +1,668 @@
+            for transition in state_transitions:
+                dest_state = transition[1]
+                global_counter_matrix += [
+                    transition[0] - global_counter_event_pos
+                ]
```
Bit of nit-picking, here and and other lines, since this is a single element that adds just one element above, IMO making compact looks better:
```suggestion
                    transition[0] - global_counter_event_pos]
```
However this suggestion is a matter of personal taste, so do as you desire

oraby reviewed (commented):
Generally LGTM, I've left few comments on the parts that I think still need addressing. But nothing major on my side.

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -271,14 +272,24 @@ def run_state_machine(self, sma):
+            # TODO: Do the BpodSystem.RefreshGUI equivalent
+            self.trial_start_timestamp = self._emulator.matrix_start_time
+        else:
+            self.trial_start_timestamp = self._bpodcom_get_trial_timestamp_start()
```
How about refactoring this whole part as follows:
```python
        if self._emulator is None:
            self._bpodcom_run_state_machine()
            if self._bpodcom_state_machine_installation_status():
                self._new_sma_sent = False
            else:
                raise BpodErrorException('Error: The last state machine sent was not acknowledged by the Bpod device.', self)
            self.trial_start_timestamp = self._bpodcom_get_trial_timestamp_start()
        else:
            self._bpodcom_run_state_machine()
            self._emulator.set_state_machine(sma)
            self._emulator.initialize()
            self._emulator.mirror_state(sma.current_state)
            # TODO: Do the BpodSystem.RefreshGUI equivalent
            self.trial_start_timestamp = self._emulator.matrix_start_time

```

oraby reviewed (commented):
One very simple change suggested, otherwise LGTM.
The only thing that I might also suggest, is to eventually squash together commits that are created then fixed in this PR into the same original initial commit.

ckaraneen commented:
Codacy analysis failures are of low severity. However, they should be dealt with in a future PR of their own.

oraby commented on pybpodapi/bpod/bpod_base.py:
```diff
@@ -271,14 +272,19 @@ def run_state_machine(self, sma):
+            self.trial_start_timestamp = self._emulator.matrix_start_time
+        else:
+            self._bpodcom_run_state_machine()
             if self._bpodcom_state_machine_installation_status():
                 self._new_sma_sent = False
             else:
                 raise BpodErrorException('Error: The last state machine sent was not acknowledged by the Bpod device.', self)
```
One last comment, we've removed line [275](https://github.com/pybpod/pybpod-api/pull/5/files#diff-b440852c8707be42a62338842bdcbe7abac80d379cc70e73fb0e3b3bc02ccc9cL275) above but it wasn't added here. These lines should be 
```suggestion
            if self._new_sma_sent:
                if self._bpodcom_state_machine_installation_status():
                    self._new_sma_sent = False
                else:
                    raise BpodErrorException('Error: The last state machine sent was not acknowledged by the Bpod device.', self)
```

oraby commented:
LGTM 👍
